src/Framework/Framework/Utils/ConcurrencyUtils.cs

```diff
@@ -0,0 +1,26 @@
+using System;
+using System.Collections.Immutable;
+using System.Threading;
+
+namespace DotVVM.Framework.Utils
+{
+    static class ConcurrencyUtils
+	{
+		public static T CasChange<T>(ref T xRef, Func<T, T> f)
+			where T : class
+		{
```

Comment:
IMO you can rewrite this method without the `goto` and `retry` label in a bit more cleaner way.

```csharp
T oldValue, newValue;

do
{
    oldValue = xRef;
    newValue = f(oldValue);
}
while (Interlocked.CompareExchange(ref xRef, newValue, oldValue) != oldValue);

return newValue;
```